src/source-dpdk.c

```diff
@@ -0,0 +1,525 @@
+{
+    int retval;
+    /* Need to be in copy mode and need to detect early release
+       where Ethernet header could not be set (and pseudo packet) */
+    if (!PKT_IS_PSEUDOPKT(p) && (p->dpdk_v.copy_mode == DPDK_COPY_MODE_TAP ||
+                                        (p->dpdk_v.copy_mode == DPDK_COPY_MODE_IPS &&
+                                                !PACKET_TEST_ACTION(p, ACTION_DROP)))) {
+        retval =
+                rte_eth_tx_burst(p->dpdk_v.out_port_id, p->dpdk_v.out_queue_id, &p->dpdk_v.mbuf, 1);
+        if (unlikely(retval < 1)) {
+            retval = rte_eth_tx_burst(
+                    p->dpdk_v.out_port_id, p->dpdk_v.out_queue_id, &p->dpdk_v.mbuf, 1);
+            if (unlikely(retval < 1)) {
+                SCLogDebug("Unable to transmit the packet on port %u queue %u",
+                        p->dpdk_v.out_port_id, p->dpdk_v.out_queue_id);
+                rte_pktmbuf_free(p->dpdk_v.mbuf);
+                p->dpdk_v.mbuf = NULL;
+            }
+        }
+    } else {
+        rte_pktmbuf_free(p->dpdk_v.mbuf);
+        p->dpdk_v.mbuf = NULL;
+    }
+
+    PacketFreeOrRelease(p);
+}
+
+/**
+ *  \brief Main DPDK reading Loop function
+ */
+TmEcode ReceiveDPDKLoop(ThreadVars *tv, void *data, void *slot)
+{
+    SCEnter();
+    Packet *p;
+    uint16_t nb_rx;
+    time_t last_dump = 0;
+    time_t current_time;
+
+    DPDKThreadVars *ptv = (DPDKThreadVars *)data;
+    TmSlot *s = (TmSlot *)slot;
+
+    ptv->slot = s->slot_next;
+
+    PacketPoolWait();
+    while (1) {
+        if (unlikely(suricata_ctl_flags != 0)) {
+            SCLogInfo("Stopping Suricata!");
+            DPDKDumpCounters(ptv);
+            break;
+        }
+
+        nb_rx = rte_eth_rx_burst(ptv->port_id, ptv->queue_id, ptv->received_mbufs, BURST_SIZE);
+        if (unlikely(nb_rx == 0)) {
+            continue;
+        }
+
+        ptv->pkts += (uint64_t)nb_rx;
+        for (uint16_t i = 0; i < nb_rx; i++) {
+            p = PacketGetFromQueueOrAlloc();
+            if (unlikely(p == NULL)) {
+                SCLogError(SC_ERR_MEM_ALLOC, "Failed to get Packet Buffer for DPDK mbuf!");
+                // exiting Suricata, we must free all remaining mbufs fetched from the NIC
+                DPDKFreeMbufArray(ptv->received_mbufs, nb_rx - i, i);
+                SCReturnInt(EXIT_FAILURE);
+            }
+            PKT_SET_SRC(p, PKT_SRC_WIRE);
+            p->datalink = LINKTYPE_ETHERNET;
+            if (ptv->checksum_mode == CHECKSUM_VALIDATION_DISABLE) {
+                p->flags |= PKT_IGNORE_CHECKSUM;
+            }
+
+            DPDKSetTimevalReal(&ptv->machine_start_time, &p->ts);
+            p->dpdk_v.mbuf = ptv->received_mbufs[i];
+            p->ReleasePacket = DPDKReleasePacket;
+            p->dpdk_v.copy_mode = ptv->copy_mode;
+            p->dpdk_v.out_port_id = ptv->out_port_id;
+            p->dpdk_v.out_queue_id = ptv->queue_id;
+
+            if (PacketSetData(p, rte_pktmbuf_mtod(p->dpdk_v.mbuf, uint8_t *),
```

Comment:
this line looks suspicious to me. I know know `rte_pktmbuf_mtod`, but are you really supposed to pass `uint8_t *` as an argument to it?

Reply:
The second argument of the function `rte_pktmbuf_mtod` causes that the result will be cast to the given type. In this case, I noticed that `PacketSetData` accepts `const uint8_t *` so therefore I am casting the result of `rte_pktmbuf_mtod` to `uint8_t *`. Do you see something wrong with it?